odoo/addons/base/models/ir_model.py

```diff
@@ -1844,6 +1844,24 @@ def group_names_with_access(self, model_name, access_mode):
         """, [lang, lang, model_name])
         return [('%s/%s' % x) if x[0] else x[1] for x in self._cr.fetchall()]
 
+    @api.model
+    @tools.ormcache('model_name', 'access_mode')
+    def _group_xmlid_with_access(self, model_name, access_mode='read'):
+        """ Return the xml ids of groups which have been granted
+            ``access_mode`` on the model ``model_name``.
+            If the tuple is empty, there is no group that can satisfy.
+            If the method returns None, it is not necessary to belong to a group to have access.
+           :rtype: tuple | None
+        """
+        assert access_mode in ('read', 'write', 'create', 'unlink'), 'Invalid access mode'
+        access = self.env['ir.model.access'].search([(f'perm_{access_mode}', '=', True), ('model_id.model', '=', model_name)])
+        group_ids = [rec.group_id.id for rec in access]
+        if not access:
+            return tuple()
+        if False in group_ids:
+            return None
+        return tuple(self.env['ir.model.data'].search([('model', '=', 'res.groups'), ('res_id', 'in', group_ids)]).mapped('complete_name'))
```

Comment:
Could be more readable with proper use of records:
```suggestion
        if not access:
            return tuple()
        if any(not acc.group_id for acc in access):
            return None
        return tuple(self.env['ir.model.data'].search([('model', '=', 'res.groups'), ('res_id', 'in', access.group_id.ids)]).mapped('complete_name'))
```